File: leetcode/algorithms/go/0005-longest-palindromic-substring.go
package leetcode

func longestPalindrome(s string) string {
	var start, length int
	var helper = func(i, j int, start, length *int) {
		for -1 < i && j < len(s) && s[i] == s[j] {
			i--
			j++
		}
		if *length < j-i-1 {
			*start = i + 1
			*length = j - i - 1
		}
	}
	for i := range s {
		helper(i, i, &start, &length)
		helper(i, i+1, &start, &length)
	}
	return s[start : start+length]
}

// func longestPalindrome(s string) string {
// 	var start, length int
// 	var helper = func(i, j, start, length int) (int, int) {
// 		for -1 < i && j < len(s) && s[i] == s[j] {
// 			i--
// 			j++
// 		}
// 		if length < j-i-1 {
// 			start = i + 1
// 			length = j - i - 1
// 		}
// 		return start, length
// 	}
// 	for i := range s {
// 		start, length = helper(i, i, start, length)
// 		start, length = helper(i, i+1, start, length)
// 	}
// 	return s[start : start+length]
// }

File: leetcode/algorithms/javascript/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    let start = 0, length = 0;
    let helper = function (i, j, start, length) {
        while (-1 < i && j < s.length && s[i] == s[j]) {
            i--;
            j++;
        }
        if (length < j - i - 1) {
            start = i + 1;
            length = j - i - 1;
        }
        return [start, length];
    };
    for (let i = 0; i < s.length; i++) {
        [start, length] = helper(i, i, start, length);
        [start, length] = helper(i, i + 1, start, length);
    }
    return s.slice(start, start + length);
};

// var longestPalindrome = function (s) {
//     let start = 0;
//     let length = 0;
//     let helper = function (i, j, start, length) {
//         while (i > -1 && j < s.length && s[i] == s[j]) {
//             i--;
//             j++;
//         }
//         if (length < j - i - 1) {
//             start = i + 1;
//             length = j - i - 1;
//         }
//         return [start, length];
//     };
//     for (let i = 0; i < s.length; i++) {
//         [start, length] = helper(i, i, start, length);
//         [start, length] = helper(i, i + 1, start, length);
//     }
//     return s.substr(start, length);
// };